### PR TITLE
DRAFT: make client more configurable, rework retry logic

### PR DESCRIPTION
**Pull request by mhoffm-aiven**
Since the purpose of this package is to be used as a library, we should make no assumptions on the environment and let the client decide how to configure this library. To this end I introduced the possibility to configure the `Client` with `Options` and deprecated the other constructors.

### CODE REVIEW

**Comment by mhoffm-aiven** [comment]
making `APIKey`, `UserAgent` and `Client` private may be a breaking change though.

**Review by ivan-savciuc — DISMISSED** [review]
I was trying to run acceptance tests, and getting lots of errors, can you please checks? 

```
    elasticsearch_acls_test.go:12: setup ElasticsearchACLs test case
2021/08/09 17:56:46 [DEBUG] POST http://127.0.0.1:51479/v1/userauth
=== RUN   TestElasticSearchACLsHandler_Get/correct
2021/08/09 17:56:46 [DEBUG] GET http://127.0.0.1:51479/v1/project/test-pr/service/test-sr/elasticsearch/acl
=== CONT  TestElasticSearchACLsHandler_Get

```

File: client.go [hunk 1]
```diff
@@ -64,87 +56,131 @@ type Client struct {
...
+	}
 }
 
-// NewTokenClient creates a new client based on a given token.
-func NewTokenClient(key string, userAgent string) (*Client, error) {
-	c := &Client{
-		APIKey:    key,
-		Client:    buildHttpClient(),
-		UserAgent: GetUserAgentOrDefault(userAgent),
+func WithRetries(retryCount uint, retryBackoff time.Duration) Option {
```
Comment (ivan-savciuc) [inline client.go:97]: It looks that we never use this function, do we need it?

File: client.go [hunk 2]
```diff
@@ -64,87 +56,131 @@ type Client struct {
...
-	if err := checkAPIResponse(bts, &r); err != nil {
-		return nil, err
+func WithMFAAuth(email, password, otp string) Option {
+	return func(cp *clientParameters) {
+		cp.authMethod = mfaAuth{email: email, otp: otp, password: password}
 	}
+}
 
-	return NewTokenClient(r.Token, userAgent)
+func WithUserAuth(email, password string) Option {
```
Comment (ivan-savciuc) [inline client.go:85]: Please add go-doc annotations to all public functions

File: client.go [hunk 3]
```diff
@@ -174,101 +210,197 @@ func (c *Client) Init() {
...
-		return responseBody, err
+	res, err = ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read request body: %w", err)
+	}
+	switch sc := rsp.StatusCode; {
+	case sc >= 200 && sc < 300:
+		// 2xx
+		return res, nil
+	case sc >= 400 && sc < 600:
```
Comment (ivan-savciuc) [inline client.go:379]: We would need to transform this to the `aiven.Error` to make `aiven.IsNotFound`() and other error methods to work correctly.

**Review by ivan-savciuc — CHANGES_REQUESTED** [review]
Nice job, foud some minor things

File: client.go (hunk 1 above)
Comment (mhoffm-aiven) [reply client.go:97]: Its intended for clients to configure retries/retry timeouts

File: client.go (hunk 2 above)
Comment (mhoffm-aiven) [reply client.go:85]: will do

**Comment by ivan-savciuc** [comment]
Closing this PR for inactivity, please reopen if still important.

**Pull request closed without merging**